File: src/problem8/Solution.java
public class Solution {
    public int myAtoi(String str) {
        if (str == null) {
            return 0;
        }
        str = str.trim();
        if (str.length() == 0) {
            return 0;
        }
        int index = 0;
        boolean isNeg = false;
        if (str.charAt(index) == '-' || str.charAt(index) == '+') {
            if (str.charAt(index) == '-') {
                isNeg = true;
            }
            index++;
        }
        long result = 0;
        while (index < str.length() && str.charAt(index) > '0' && str.charAt(index) < '9') {
            result = result * 10 + (int) str.charAt(index) - 48;
            if (result > Integer.MAX_VALUE) {
                if (isNeg) {
                    return Integer.MIN_VALUE;
                } else {
                    return Integer.MAX_VALUE;
                }
            }
            index++;
        }
        return isNeg ? (int) -result : (int) result;
    }
}

File: src/problem8/Solution.cpp
class Solution {
public:
    int myAtoi(string str) {
        if (str.empty())    return 0;
        int index = 0;
        while (str[index] == ' ')   ++index;
        long long result = 0;
        int sign = 1;
        if (str[index] == '-') {
            ++index;
            sign = -1;
        } else if (str[index] == '+') {
            ++index;
            sign = 1;
        }
        for (int i = index; i < str.size(); ++i) {
            if (str[i] >= '0' && str[i] <= '9') {
                result = result * 10 + (int)(str[i]-'0');
                if (result > INT_MAX) {
                    return sign==1 ? INT_MAX : INT_MIN;
                }
            } else {
                break;
            }
        }
        return sign == 1 ? result : -result;
    }
};